Remove old CTC decoder (Fixes #1675)

reuben opened the pull request:
- [x] Remove old decoder code and build system integration
- [x] Remove references in code and tests
- [x] Update documentation

reuben commented:
@lissyx I'm having trouble getting the ctcdecode package to build on ARM64, getting relocation errors in libm.a when doing the final linking of the Python extension. I see that in native_client/BUILD we link against libstdc++.a, is that a requirement on our ARM platforms? The strange thing is, it's building fine on ARMv7...

lissyx commented:
> @lissyx I'm having trouble getting the ctcdecode package to build on ARM64, getting relocation errors in libm.a when doing the final linking of the Python extension. I see that in native_client/BUILD we link against libstdc++.a, is that a requirement on our ARM platforms? The strange thing is, it's building fine on ARMv7...

They should share the same linking strategy, but maybe you lack some `-PIE` or something like that?

lissyx commented on taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml:
```diff
@@ -1,7 +1,7 @@
   template_file: test-linux-opt-base.tyml
   dependencies:
-    - "linux-amd64-ctc-opt"
+    - "linux-amd64-cpu-opt"
```
I don't think it's a good idea, completion time of `linux-amd64-cpu-opt` is much higher, it's going to delay everything behind, we should re-use `linux-amd64-ctc-opt` to build the new python wheel instead

reuben replied:
The problem is that it's not a new Python wheel, it's several Python wheels, one for each architecture/Python version, so I tried to leverage the existing infrastructure for CPU builds.

lissyx replied:
Sure, but it's still better IMHO to do that outside, we can easily hack around `do_deepspeech_python_build`

reuben commented:
Initially the compiler error mentioned -fPIC, but compiling with that flag made no difference, it still fails with relocation issues in libm.a.

reuben commented:
I think the problem is that definitions.mk doesn't cover all that's needed to get a build working on ARM, as it's missing some of the stuff that was added in the tensorflow CROSSTOOL file like pointing to a different linker.

lissyx commented:
> I think the problem is that definitions.mk doesn't cover all that's needed to get a build working on ARM, as it's missing some of the stuff that was added in the tensorflow CROSSTOOL file like pointing to a different linker.

It's more simply just unable to find the proper `libm` because of absolute paths in ld scripts. Quick hack that works for me:
```
diff --git a/native_client/ctcdecode/Makefile b/native_client/ctcdecode/Makefile
index 14cc591..39c5e63 100644
--- a/native_client/ctcdecode/Makefile
+++ b/native_client/ctcdecode/Makefile
@@ -4,6 +4,8 @@ include ../definitions.mk

 NUM_PROCESSES ?= 1

+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
+
 all: bindings

 clean:
@@ -16,4 +18,4 @@ bindings:
        AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS) $(CXXFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py build_ext --num_processes $(NUM_PROCESSES) $(PYTHON_PLATFORM_NAME) $(SETUP_FLAGS)
        find temp_build -type f -name "*.o" -delete
        AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS) $(CXXFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py bdist_wheel --num_processes $(NUM_PROCESSES) $(PYTHON_PLATFORM_NAME) $(SETUP_FLAGS)
-       rm -rf temp_build
\ No newline at end of file
+       rm -rf temp_build
diff --git a/native_client/ctcdecode/setup.py b/native_client/ctcdecode/setup.py
index e6b3855..ec31f9b 100644
--- a/native_client/ctcdecode/setup.py
+++ b/native_client/ctcdecode/setup.py
@@ -89,10 +89,6 @@ LIBS = []
 ARGS = ['-O3', '-DNDEBUG', '-DKENLM_MAX_ORDER=6', '-std=c++11',
         '-Wno-unused-local-typedef', '-Wno-sign-compare']

-if os.environ['TARGET'] == 'rpi3-armv8':
-    ARGS.append('-static-libstdc++')
-
-
 decoder_module = Extension(
     name='ds_ctcdecoder._swigwrapper',
     sources=['swigwrapper.i'] + FILES + glob.glob('*.cpp'),
```

reuben commented:
We have liftoff! Thanks @lissyx!

lissyx commented on native_client/ctcdecode/Makefile:
```diff
@@ -4,6 +4,11 @@ include ../definitions.mk
 
+# ARM64 can't find the proper libm.so without this
+ifeq ($(TARGET),rpi3-armv8)
+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
```
Even if it's not very nice, we should avoid adding more unniceness with hard-coding the version. `libm.so.6` should be a good stable symlink

lissyx commented on native_client/ctcdecode/scorer.cpp:
```diff
@@ -43,6 +43,16 @@ Scorer::Scorer(double alpha,
+               const char* alphabet_config_path)
+  : Scorer(alpha, beta, lm_path, trie_path, Alphabet(alphabet_config_path))
+{
+
```
nit: no empty line?

lissyx commented on native_client/ctcdecode/scorer.cpp:
```diff
@@ -114,7 +124,7 @@ void Scorer::save_dictionary(const std::string& path) {
 
 double Scorer::get_log_cond_prob(const std::vector<std::string>& words) {
-  double cond_prob;
+  double cond_prob = OOV_SCORE;
```
I dont see this `OOV_SCORE`, where is it coming from?

kdavis-mozilla replied:
It's in `scorer.h`

lissyx commented on taskcluster/decoder-build.sh:
```diff
@@ -4,10 +4,7 @@ set -xe
 source $(dirname "$0")/../tc-tests-utils.sh
 
-source ${DS_ROOT_TASK}/DeepSpeech/tf/tc-vars.sh
+export OS=$(uname)
```
There's already this in `tc-tests-utils.sh`

reuben replied:
Maybe it's the lack of `export` there, but without this the files downstream were picking up the `export OS` in tf-vars.sh, not `tc-tests-utils.sh`. In any case, I'll add the `export` in `tc-tests-utils.sh` instead.

tilmankamp commented on util/coordinator.py:
```diff
@@ -0,0 +1,706 @@
+            result = pickle.loads(result)
+        return result
+
+class GlobalConfig:
```
This should not live here in the coordinator. You could also somehow join it with the flags module.

reuben replied:
There's a circular dependency between the globals and the coordinator (the COORD global) and IMO breaking it by separating just that global initialization out of `initialize_globals` would be worse than having these live in the same place. What do you think?

tilmankamp replied:
I'd initalize the `COORD` global variable in `DeepSpeech.py` (moving the block with comment `# Initializing and starting the training coordinator` back there). Then there should be a clean separation of concerns again.

tilmankamp commented on util/coordinator.py:
```diff
@@ -0,0 +1,706 @@
+
+C = GlobalConfig()
+
+def initialize_globals():
```
This should not live here in the coordinator.

tilmankamp commented on util/coordinator.py:
```diff
@@ -0,0 +1,706 @@
+from __future__ import absolute_import, division, print_function
```
It's good to have this in an extra file, as this should go away soon. We should soon move to an in-graph replica scheme for training.

tilmankamp commented on DeepSpeech.py:
```diff
@@ -8,356 +8,24 @@
 log_level_index = sys.argv.index('--log_level') + 1 if '--log_level' in sys.argv else 0
 os.environ['TF_CPP_MIN_LOG_LEVEL'] = sys.argv[log_level_index] if log_level_index > 0 and log_level_index < len(sys.argv) else '3'
 
-import datetime
```
Modularizing this file is a great thing. But why not doing it in a separate PR? I never got a similar refactoring merged due to an overloaded PR.

reuben replied:
That's why I split this into commits, so you can look at each commit individually. I can also make separate PRs for the commits, but that will require making each PR buildable/passing tests, which will be a lot more work.

tilmankamp replied:
I know - that's also the reason why I don't insist in doing it. :-) Doing this after the fact is much harder and a big waste of time.

tilmankamp commented on DeepSpeech.py:
```diff
@@ -1716,7 +586,7 @@ def update_progressbar(set_name):
 
                         log_debug('Starting batch...')
                         # Compute the batch
-                        _, current_step, batch_loss, batch_report, step_summary = session.run([train_op, global_step, loss, report_params, step_summaries_op], **extra_params)
```
Does this mean we don't compute training and validation WER reports anymore?

lissyx commented on taskcluster/host-build.sh:
```diff
@@ -21,6 +21,7 @@ do_deepspeech_binary_build
 
 do_deepspeech_python_build
 
+do_deepspeech_decoder_build
```
Why do we need this one ? In the past, there was a `decoder-build.sh` call to build the `libctc_decoder_with_kenlm.so` for packaging it, but this one does not get packaged, right?

reuben replied:
Good point, I don't think this is needed.

lissyx commented on .install:
```diff
@@ -7,4 +7,10 @@ pip install tensorflow-gpu==1.12.0rc2
+# Install ds_ctcdecoder package from TaskCluster
+VERSION=$(python -c 'import pkg_resources; print(pkg_resources.safe_version(open("VERSION").read()))')
+PYVER=$(python -c 'import sys; print("cp{0}{1}-cp{0}{1}m".format(sys.version_info.major, sys.version_info.minor))')
+python3 util/taskcluster.py --arch cpu --target ../tmp --artifact "ds_ctcdecoder-${VERSION}-${PYVER}-manylinux1_x86_64.whl"
```
this will install from `master` then ?

reuben commented:
In 38b5447 I've updated the docs and added some convenience code in `util/taskcluster.py` to get the URL of the `ds_ctcdecoder` package for easy installation.

kdavis-mozilla commented on evaluate.py:
```diff
@@ -15,8 +15,10 @@
-from DeepSpeech import initialize_globals, create_flags, log_debug, log_info, log_warn, log_error, create_inference_graph
-from multiprocessing import Pool
+from util.flags import create_flags
+from util.coordinator import C, initialize_globals
```
Again the naming of `C` seems too generic.

kdavis-mozilla commented on evaluate.py:
```diff
@@ -221,7 +206,38 @@ def create_windows(features):
+        by="features_len",
+        ascending=False)
+
+    from DeepSpeech import create_inference_graph
```
Why import here and not at file's top?

reuben replied:
That would make it a circular dependency as DeepSpeech.py imports evaluate.py.

kdavis-mozilla commented on util/taskcluster.py:
```diff
@@ -74,9 +78,55 @@ def maybe_download_tc_bin(**kwargs):
+        if plat == 'darwin':
+            plat = 'macosx_10_10'
+
+        from pkg_resources import parse_version
```
Just wondering is there is a reason to use local imports instead of globals.

kdavis-mozilla commented on util/coordinator.py:
```diff
@@ -186,7 +184,7 @@ def done(self):
                 # if the job was for validation dataset then append it to the COORD's _loss for early stop verification
                 if (FLAGS.early_stop is True) and (self.set_name == 'dev'):
-                    COORD._dev_losses.append(self.loss)
+                    self.coord_dev_losses.append(self.loss)
```
This should be
```python
self.coord._dev_losses.append(self.loss)
```

kdavis-mozilla commented on util/coordinator.py:
```diff
@@ -415,15 +417,15 @@ def start(self):
             log_debug('Starting coordinator...')
             self._thread = Thread(target=self._httpd.serve_forever)
-            self._thread.daemon = True
+            # self._thread.daemon = True
```
Out of curiosity, why did this change?

reuben replied:
Leftover from debugging, will remove this.

reuben commented on DeepSpeech.py:
```diff
@@ -8,356 +8,24 @@
 from util.feeding import DataSet, ModelFeeder
+from util.logging import *
+from util.flags import create_flags, FLAGS
+from util.coordinator import C, initialize_globals
```
The reason for the short name is to make it easier to transition our generous use of globals to more reasonable solutions. A long name would screw up code style across the project. But if you feel strongly about this I'll rename it to Config or something like that.

reuben commented on DeepSpeech.py:
```diff
@@ -1948,8 +824,10 @@ def __init__(self):
+    import numpy as np
+
+    with tf.Session(config=C.session_config) as session:
+        inputs, outputs, _ = create_inference_graph(batch_size=1, n_steps=-1, use_new_decoder=True)
```
This was already unused before this PR. Removed.

reuben commented:
Everything should be addressed now.

kdavis-mozilla commented:
LGTM

lissyx commented on native_client/ctcdecode/Makefile:
```diff
@@ -6,7 +6,7 @@ NUM_PROCESSES ?= 1
 # ARM64 can't find the proper libm.so without this
 ifeq ($(TARGET),rpi3-armv8)
-LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm.so.6
```
:+1:

lissyx commented on util/taskcluster.py:
```diff
@@ -59,13 +60,16 @@ def maybe_download_tc_bin(**kwargs):
                         help='Where to put the native client binary files')
-    parser.add_argument('--arch', required=False, default='cpu',
+    parser.add_argument('--arch', required=False,
                         help='Which architecture to download binaries for. "arm" for ARM 7 (32-bit), "gpu" for CUDA enabled x86_64 binaries, "cpu" for CPU-only x86_64 binaries, "osx" for CPU-only x86_64 OSX binaries. Optional ("cpu" by default)')
```
nit: also mention arm64 here, since we its default value

reuben commented:
Yes.

-- reuben

> On 15 Nov 2018, at 03:17, Josh Meyer <notifications@github.com> wrote:
> 
> @reuben, does this PR mean we don't have to generate the old trie model (i.e. from v0.3.0) for training and new trie model for testing?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

JRMeyer commented:
@reuben I'm loading in an older checkpoint (v0.3.0), and I get the following error:

```
KeyError: 'CTCBeamSearchDecoderWithLM'
````

After some searching, I think this is related to a missing `libctc_decoder_with_kenlm.so` file (c.f. #1306, #950)

 The following is the script I'm running to load the checkpoint:


```
import tensorflow as tf
from tensorflow.python.tools import freeze_graph
from ds_ctcdecoder import ctc_beam_search_decoder, Scorer

sess = tf.Session()
sess.run(tf.global_variables_initializer())
 
tf.contrib.rnn

tf.load_op_library('../tmp/native_client/libdeepspeech.so')

new_saver = tf.train.import_meta_graph('../v030/ckpt/model.v0.2.0.meta')
new_saver.restore(sess, '../v030/ckpt/model.v0.2.0')
```

Is this error a by-product of this PR? Are the older checkpoints on the release page not forwards compatible after this PR?

-josh

JRMeyer commented:
Strange, but good news.
